File: src/ElroukenAljamil.Application/Services/WorkflowService.cs
using System.Text.Json;
using ElroukenAljamil.Application.DTOs;
using ElroukenAljamil.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace ElroukenAljamil.Application.Services;

public class WorkflowService
{
    private readonly AppDbContext _context;

    public WorkflowService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<WorkflowDto?> GetWorkflowByCategoryId(int categoryId)
    {
        var workflow = await _context.DepositWorkflows
            .Include(w => w.Steps.Where(s => s.IsActive))
                .ThenInclude(s => s.Fields.Where(f => f.IsActive))
            .Where(w => w.CategoryId == categoryId && w.IsActive)
            .FirstOrDefaultAsync();

        if (workflow == null)
        {
            // Try parent category
            var category = await _context.Categories.FindAsync(categoryId);
            if (category?.ParentCategoryId != null)
            {
                workflow = await _context.DepositWorkflows
                    .Include(w => w.Steps.Where(s => s.IsActive))
                        .ThenInclude(s => s.Fields.Where(f => f.IsActive))
                    .Where(w => w.CategoryId == category.ParentCategoryId && w.IsActive)
                    .FirstOrDefaultAsync();
            }
        }

        if (workflow == null) return null;

        return MapToDto(workflow);
    }

    public async Task<List<WorkflowDto>> GetAllWorkflows()
    {
        var workflows = await _context.DepositWorkflows
            .Include(w => w.Category)
            .Include(w => w.Steps.Where(s => s.IsActive))
                .ThenInclude(s => s.Fields.Where(f => f.IsActive))
            .Where(w => w.IsActive)
            .ToListAsync();

        return workflows.Select(MapToDto).ToList();
    }

    private WorkflowDto MapToDto(Domain.Entities.DepositWorkflow workflow)
    {
        return new WorkflowDto
        {
            Id = workflow.Id,
            CategoryId = workflow.CategoryId,
            Name = workflow.Name,
            Description = workflow.Description,
            Steps = workflow.Steps
                .OrderBy(s => s.StepOrder)
                .Select(s => new WorkflowStepDto
                {
                    Id = s.Id,
                    StepOrder = s.StepOrder,
                    Title = s.Title,
                    Subtitle = s.Subtitle,
                    StepKey = s.StepKey,
                    IsRequired = s.IsRequired,
                    Fields = s.Fields
                        .OrderBy(f => f.DisplayOrder)
                        .Select(f => new StepFieldDto
                        {
                            Id = f.Id,
                            FieldKey = f.FieldKey,
                            Label = f.Label,
                            FieldType = f.FieldType,
                            Placeholder = f.Placeholder,
                            Options = ParseOptions(f.Options),
                            DefaultValue = f.DefaultValue,
                            Suffix = f.Suffix,
                            HelperText = f.HelperText,
                            IsRequired = f.IsRequired,
                            DisplayOrder = f.DisplayOrder,
                            MaxLength = f.MaxLength
                        }).ToList()
                }).ToList()
        };
    }

    private List<FieldOptionDto> ParseOptions(string optionsJson)
    {
        if (string.IsNullOrWhiteSpace(optionsJson)) return new();
        try
        {
            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            return JsonSerializer.Deserialize<List<FieldOptionDto>>(optionsJson, options) ?? new();
        }
        catch
        {
            return new();
        }
    }
}

File: elrouken-aljamil-client/src/app/services/workflow.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface FieldOption {
  value: string;
  label: string;
}

export interface StepField {
  id: number;
  fieldKey: string;
  label: string;
  fieldType: string;
  placeholder: string;
  options: FieldOption[];
  defaultValue: string;
  suffix: string;
  helperText: string;
  isRequired: boolean;
  displayOrder: number;
  maxLength: number | null;
  visibilityCondition: string;
}

export interface WorkflowStep {
  id: number;
  stepOrder: number;
  title: string;
  subtitle: string;
  stepKey: string;
  isRequired: boolean;
  fields: StepField[];
}

export interface Workflow {
  id: number;
  categoryId: number;
  name: string;
  description: string;
  steps: WorkflowStep[];
}

@Injectable({ providedIn: 'root' })
export class WorkflowService {
  private apiUrl = 'https://localhost:7283/api/workflow';

  constructor(private http: HttpClient) {}

  getWorkflowByCategory(categoryId: number): Observable<Workflow> {
    return this.http.get<Workflow>(`${this.apiUrl}/${categoryId}`);
  }

  getAllWorkflows(): Observable<Workflow[]> {
    return this.http.get<Workflow[]>(this.apiUrl);
  }
}
